### PR TITLE
backupccl,server: add DownloadSpans client side API

### PR DESCRIPTION
**Pull request by adityamaru**
This change adds an endpoint to the status
server, tenant connector and system status
server to instruct pebble to download a span.

The endpoint fans out to all nodes in the cluster, and instructs every store on each node to download the external files that intersect with the span.

This change also wires up online restore to issue
these download requests once the linking phase is complete. Currently, we issue download requests for each index span in the backup but going forward we will want to define prioritization heuristics for what spans to send.

Epic: none
Release note: None

### CODE REVIEW

File: pkg/storage/pebble.go [hunk 1]
```diff
@@ -619,6 +620,9 @@ func DefaultPebbleOptions() *pebble.Options {
 		FormatMajorVersion:          MinimumSupportedFormatVersion,
 	}
 	opts.Experimental.L0CompactionConcurrency = l0SubLevelCompactionConcurrency
+	opts.Experimental.IngestSplit = func() bool {
```
Comment (msbutler) [inline pkg/storage/pebble.go:623]: this should probably be in its own commit? perhaps not in this pr all together?

**Review by msbutler — COMMENTED** [review]
looks good! just one blocking comment on enabling ingestSplit in pebble, by default.

File: build/bazelutil/distdir_files.bzl [hunk 2]
```diff
@@ -321,7 +321,6 @@ DISTDIR_FILES = {
     "https://storage.googleapis.com/cockroach-godeps/gomod/github.com/cockroachdb/go-test-teamcity/com_github_cockroachdb_go_test_teamcity-v0.0.0-20191211140407-cff980ad0a55.zip": "bac30148e525b79d004da84d16453ddd2d5cd20528e9187f1d7dac708335674b",
     "https://storage.googleapis.com/cockroach-godeps/gomod/github.com/cockroachdb/gostdlib/com_github_cockroachdb_gostdlib-v1.19.0.zip": "c4d516bcfe8c07b6fc09b8a9a07a95065b36c2855627cb3514e40c98f872b69e",
     "https://storage.googleapis.com/cockroach-godeps/gomod/github.com/cockroachdb/logtags/com_github_cockroachdb_logtags-v0.0.0-20230118201751-21c54148d20b.zip": "ca7776f47e5fecb4c495490a679036bfc29d95bd7625290cfdb9abb0baf97476",
-    "https://storage.googleapis.com/cockroach-godeps/gomod/github.com/cockroachdb/pebble/com_github_cockroachdb_pebble-v0.0.0-20231114155657-6682fd5e8763.zip": "496e6b57b21b0e08c7e8de41a597f629fa1d1e0ac21c3b1cfd3692ac88ec193c",
```
[the remark this answers is not in the document]
Comment (adityamaru) [reply build/bazelutil/distdir_files.bzl:324]: indeed

File: pkg/storage/pebble.go (hunk 1 above)
Comment (adityamaru) [reply pkg/storage/pebble.go:623]: nice catch, definitely didn't mean to make this call for storage.

File: pkg/ccl/backupccl/restore_online.go [hunk 3]
```diff
@@ -230,3 +242,171 @@ func checkRewritesAreNoops(rewrites jobspb.DescRewriteMap) error {
...
+			lastProgressUpdate = timeutil.Now()
+		}
+	}
+}
+
+func (r *restoreResumer) doDownloadFiles(ctx context.Context, execCtx sql.JobExecContext) error {
+	details := r.job.Details().(jobspb.RestoreDetails)
+
+	grp := ctxgroup.WithContext(ctx)
+	downloadSpansCh := make(chan roachpb.Span, len(details.DownloadSpans))
```
[the remark this answers is not in the document]
Comment (adityamaru) [reply pkg/ccl/backupccl/restore_online.go:389]: I didn't want this channel to be a bottleneck. We could have multiple workers pulling off this channel and issuing download requests, and we should just fill the channel as quickly as we can to avoid any of those workers waiting for work.

File: pkg/ccl/backupccl/restore_online.go [hunk 4]
```diff
@@ -230,3 +242,171 @@ func checkRewritesAreNoops(rewrites jobspb.DescRewriteMap) error {
...
+			var resp *serverpb.DownloadSpanResponse
+			var err error
+			if resp, err = execCtx.ExecCfg().TenantStatusServer.DownloadSpan(ctx, &serverpb.DownloadSpanRequest{
+				Span: sp,
+			}); err != nil {
+				return err
+			}
+			if len(resp.ErrorsByNodeID) > 0 {
+				return errors.Newf("failed to download span %s on all nodes: %v", sp, resp.ErrorsByNodeID)
+			}
```
[the remark this answers is not in the document]
Comment (adityamaru) [reply pkg/ccl/backupccl/restore_online.go:317]: as in, do something more than fail the job right? I agree

File: pkg/kv/kvclient/kvtenant/connector.go [hunk 5]
```diff
@@ -880,6 +880,24 @@ func (c *connector) HotRangesV2(
 	return resp, nil
 }
 
+// DownloadSpan implements the serverpb.TenantStatusServer interface
+func (c *connector) DownloadSpan(
+	ctx context.Context, req *serverpb.DownloadSpanRequest,
+) (*serverpb.DownloadSpanResponse, error) {
+	if !roachpb.IsSystemTenantID(c.tenantID.InternalValue) {
+		return nil, status.Errorf(codes.PermissionDenied, "only the system tenant can issue download span requests")
+	}
```
[the remark this answers is not in the document]
Comment (adityamaru) [reply pkg/kv/kvclient/kvtenant/connector.go:889]: sorry i dont think i follow, is the suggestion to do `if c.tenantID.InternalValue != roachpb.SystemTenantID` instead of using the helper `if !roachpb.IsSystemTenantID(c.tenantID.InternalValue)`?

File: pkg/ccl/backupccl/restore_online.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (msbutler) [reply pkg/ccl/backupccl/restore_online.go:389]: I see. I don't have strong opinions here, so I'm fine to keep it as is.

File: pkg/ccl/backupccl/restore_online.go [hunk 6]
```diff
@@ -230,3 +242,171 @@ func checkRewritesAreNoops(rewrites jobspb.DescRewriteMap) error {
...
+		return 0.0
+	}); err != nil {
+		return 0, err
+	}
+
+	if err := r.job.NoTxn().RunningStatus(ctx, func(_ context.Context, _ jobspb.Details) (jobs.RunningStatus, error) {
+		return jobs.RunningStatus(fmt.Sprintf("Downloading %s of restored data...", sz(total))), nil
+	}); err != nil {
+		return 0, errors.Wrapf(err, "failed to update running status of job %d", errors.Safe(r.job.ID()))
+	}
```
[the remark this answers is not in the document]
Comment (adityamaru) [reply pkg/ccl/backupccl/restore_online.go]: changed it to use `Update` and do both in the same call, check it out!

**Review by msbutler — APPROVED** [review]

**Review by dt — APPROVED** [review]

**Comment by adityamaru** [comment]
TFTR!!

bors r+
